reconcile/test/test_utils_state.py

```diff
@@ -0,0 +1,56 @@
+import pytest
+import boto3
+from moto import mock_s3
+from reconcile.utils.state import State
+
+
+@pytest.fixture
+def accounts():
+    """Account name is needed to instantiate a State"""
+    return [{'name': 'some-account'}]
+
+
+@mock_s3
+def test_ls(accounts, monkeypatch, mocker):
+    monkeypatch.setenv('AWS_ACCESS_KEY_ID', 'testing')
+    monkeypatch.setenv('AWS_SECRET_ACCESS_KEY', 'testing')
+    monkeypatch.setenv('AWS_SECURITY_TOKEN', 'testing')
+    monkeypatch.setenv('AWS_SESSION_TOKEN', 'testing')
+    monkeypatch.setenv('APP_INTERFACE_STATE_BUCKET', 'some-bucket')
+    monkeypatch.setenv('APP_INTERFACE_STATE_BUCKET_ACCOUNT', 'some-account')
+
+    s3_client = boto3.client('s3', region_name='us-east-1')
+    s3_client.create_bucket(Bucket='some-bucket')
+
+    s3_client.put_object(Bucket='some-bucket',
+                         Key='state/integration-name/some-file-1',
+                         Body='test')
+
+    """Creating some-file-2 to identify when two or more integrations have
+    similar names"""
+    s3_client.put_object(Bucket='some-bucket',
+                         Key='state/integration-name-2/some-file-2',
+                         Body='test')
+
+    class MockAWSAPi:
+        def __init__(self, thread_pool_size, accounts,
+                     settings=None):
+            pass
+
+        @staticmethod
+        def get_session(account):
+            class Session:
+                @staticmethod
+                def client(type):
+                    return s3_client
+            return Session()
+
+    mocker.patch('reconcile.utils.state.AWSApi', MockAWSAPi)
+
+    state = State('integration-name', accounts)
+
+    objects = state.ls()
+
+    expected = ['/some-file-1']
```

Comment:
The keys having leading forward slashes is odd to me.  It seems that all the other methods take a bare key without the forward slash?  Also, `get_all` is removing that forward slash?  I'd expect to be able to do something like:

```
state_keys = state.ls()

for key in state_keys:
    state.get(key)
```

From quickly reading the code, it seems that you'd need to account for the slash to make that work.  I can see the couple integrations using `ls()` dealing with this, but wonder if this makes sense or whether it should be addressed in a separate PR?

Also, `qontract_cli.py` has some interesting usage of this:

```
    keys = state.ls()
    # if 'integration' is defined, the 0th token is empty
    table_content = [
        {'integration': k.split('/')[0] or integration,
         'key': '/'.join(k.split('/')[1:])}
        for k in keys]
```

Did you already come across that?

Reply:
> Did you already come across that?

Yes, that was why I commented [here](https://github.com/app-sre/qontract-reconcile/pull/1848#discussion_r709484431)

IMO a refactoring that has the potential to change that much files should be in another PR.

Reply:
I agree and that's why I mentioned the separate PR.  This class doesn't seem to be widely used yet, so making a breaking change (returning without forward slash) would only require fixes in a couple of places, rather than later it could be many.  I think it's worth a tracking ticket in any case?

Reply:
Done: https://issues.redhat.com/browse/APPSRE-3781